draft-rescorla-tls-esni.md

```diff
@@ -247,22 +243,28 @@ The SNI encryption key is computed from the DH shared secret Z as
 follows:
 
 ~~~~
-   Z_extracted = HKDF-Extract(EncryptedSNI.nonce, Z)
-   K_sni = HKDF-Expand-Label(Z_extracted, "encrypted-sni", ClientHello.Random, L)
-
-   Where L is the key length associated with the cipher suite.
+   Z_extracted = HKDF-Extract(0, Z)
+   key = HKDF-Expand-Label(Z_extracted, "esni key", ClientHello.Random, key_length)
+   iv = HKDF-Expand-Label(Z_extracted, "esni iv", ClientHello.Random, iv_length)
 ~~~~
 
-The EncryptedSNI.encrypted_sni value is then computed by:
+The EncryptedSNI.encrypted_sni value is then computed using the usual
+TLS 1.3 AEAD:
 
 ~~~~
-    encrypted_sni = AEAD-Encrypt(K_sni, 0, "", ServerNameList)
+    encrypted_sni = AEAD-Encrypt(key, iv, "", ServerNameList)
 ~~~~
 
-[[OPEN ISSUE: This is a strawman construction. We do want a
-nonce to avoid situations where the server somehow reuses
-a key, but exactly how we mix it in is TBD. Maybe in both
-places?]]
+
+Note: future extensions may end up reusing the server's ESNIKeyShare
+for other purposes within the same message (e.g., encrypting other
+values). Those usages MUST have their own HKDF labels to avoid
+reuse.
+
+[[OPEN ISSUE: If in future you were to reuse these keys for
+0-RTT priming, then you would have to worry about potentially
+expanding twice of Z_extracted. We should think about how
+to harmonize these.]]
```

Comment:
This is an implementation detail, I think. If a client chooses to expand twice when it's not needed then that's their fault. We could simply state (in the other draft, or here) that "if clients have computed Z for other purposes, an additional extraction is not needed."

Reply:
Sorry, I'm not worried about the implementation, but about whether we're guaranteeing key separation